algorithms/src/snark/groth16/mod.rs

```diff
@@ -135,19 +135,19 @@ impl<E: PairingEngine> Proof<E> {
 
     /// Deserialize a proof from compressed or uncompressed bytes.
     pub fn read<R: Read>(mut reader: R) -> IoResult<Self> {
-        // Construct the compressed reader.
         let compressed_proof_size = Self::compressed_proof_size()?;
-        let mut proof_reader = vec![0u8; compressed_proof_size];
-        reader.read_exact(&mut proof_reader)?;
+        let uncompressed_proof_size = Self::uncompressed_proof_size()?;
+
+        // Construct the compressed reader.
+        let mut proof_reader = vec![0u8; uncompressed_proof_size];
```

Comment:
Just to check, is it more efficient/cheaper to allocate double the expected size first, rather than resizing after the compressed case is checked?

Reply:
Yes - it is faster, and for hot enough functions the difference in performance is detectable.